fullstorydev/hauser
# Convert to UTC before saving sync point to column with data type timestamp without timezone

This PR converts the bundle end time to UTC before persisting it to the sync table on Redshift. We have to do this because the type of the column on redshift is timestamp without timezone. 

*High level flow in Hauser*
1) Fetch the latest `bundleEndTime` from the sync table
2) Use the `bundleEndTime`  as the start param when calling the `export/list` end point. This returns all the bundles starting from the bundle that contains the provided start time.
3) Iterate over all the bundles returned from `export/list` 
- fetch data via export/get
- write bundle data to the export table
- after bundle is successfully persisted to the export table, add a record for the bundle to the sync table. This so we can start from the last bundle we haven't processed yet.

*The Problem*
Since the column on redshift doesn't have timezone, when we read the column we assume the time has the local timezone. For our customers in the US, it shifts the time back by a few hours. When this time was sent as the start param to the `export/list` endpoint it returned the last available bundle over and over again.

After converting to UTC, before persisting we accurately interpret the 'bundleEndTime' column in the sync table, and accurately determine there are no more bundles to process instead of repeatedly querying the last bundle repeatedly until a new bundle becomes available. 

I should also note duplicate records are not found in the export. We have a mechanism in place that protects us from cases where the export table is updated, but the sync table fails to update. This mechanism has been preventing us from writing duplicates in this situation as well.

*EDIT (2019-09-06)*: I may have spoken too soon about duplicate records. If someone has a bundle size that is greater than the time by which the shift occurs, then they will have duplicate records because the RemoveOrphanedRecords fn will only remove a subset of the previous bundle before re-processing and inserting the entire bundle again.
